include/clap/factory/draft/preset-discovery.h

```diff
@@ -0,0 +1,321 @@
+   //
+   // name is mandatory, it is the preset's name.
+   //
+   // If the preset file is a preset container then subpath and load_key are mandatory,
+   // otherwise they must be null.
+   //
+   // The subpath defines a human friendly path to the preset in the container file.
+   // It must be unique within the container file.
+   //
+   // The load_key is a machine friendly string used to load the preset inside the container via a
+   // the preset-load plug-in extension. The load_key can also just be the subpath if that's what
+   // the plugin wants but it could also be some other unique id like a database primary key or a
+   // binary offset. It's use is entirely up to the plug-in.
+   //
+   // If the function returns false, the the provider must stop calling back into the receiver.
+   bool(CLAP_ABI *begin_preset)(const struct clap_preset_discovery_metadata_receiver *receiver,
+                                const char                                           *name,
+                                const char                                           *subpath,
+                                const char                                           *load_key);
+
+   // Adds a plug-in id that this preset can be used with.
+   // plugin_abi: 0 for CLAP
+   void(CLAP_ABI *add_plugin_id)(const struct clap_preset_discovery_metadata_receiver *receiver,
+                                 uint32_t                                              plugin_abi,
+                                 const char                                           *plugin_id);
+
+   // Sets the sound pack to which the preset belongs to.
+   void(CLAP_ABI *set_soundpack_id)(const struct clap_preset_discovery_metadata_receiver *receiver,
+                                    const char *soundpack_id);
+
+   // Sets the flags, see clap_preset_discovery_flags.
+   // If unset, they are then inherited from the location.
+   void(CLAP_ABI *set_flags)(const struct clap_preset_discovery_metadata_receiver *receiver,
+                             uint32_t                                              flags);
+
+   // Adds a creator name for the preset.
+   void(CLAP_ABI *add_creator)(const struct clap_preset_discovery_metadata_receiver *receiver,
+                               const char                                           *creator);
+
+   // Sets a description of the preset.
+   void(CLAP_ABI *set_description)(const struct clap_preset_discovery_metadata_receiver *receiver,
+                                   const char *description);
+
+   // Sets the creation time and last modification time of the preset.
+   // The timestamps are in seconds since UNIX EPOCH, see C's time_t time(time_t *).
+   // If one of the time isn't known, then set it to 0.
+   // If this function is not called, then the indexer may look at the file's creation and
+   // modification time.
+   void(CLAP_ABI *set_timestamps)(const struct clap_preset_discovery_metadata_receiver *receiver,
```

Comment:
Thinking about preset versioning/variations (see also my loose thoughts/ideas #284 ): Could we allow multiple presets with same subpath and creation timestamps but different modification timestamps? This might be helpful for managing different versions/variations of the same preset.

Reply:
It could then be up to the host to decide whether to show only the newest version of a preset, list the preset variations next to each other or to offer a separate selector for the preset versions (like the selector when someone edited a github comment ;) ).

Reply:
It'd be too complicated.
That would make the API more complicated for a feature very few will use.
If you make multiple variations, then I'd suggest to use similar preset name and put the variation at the end.
Though I suggest that when making a sound pack, to just keep the best variation so it is then less content to browse for the user. The user's time is precious and it is better to not make him try 5 times the same preset, also to not unnecessarily clutter the preset browser.